File: SampleApp/basefiles/src/main/java/com/rp/basefiles/BaseAdapterPresenter.java
package com.rp.basefiles;

import android.support.v7.recyclerview.extensions.AsyncListDiffer;
import android.support.v7.util.DiffUtil;

import com.rp.util.adapter.RAdapterAsyncDiffCallback;
import com.rp.util.adapter.RAdapterPayloadWatcher;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by rahul on 20/1/18.
 */

public abstract class BaseAdapterPresenter<V extends IBaseHolderView, E>
        implements IBaseAdapterPresenter<E> {

    private V baseHolderView;

    private AsyncListDiffer<E> mDiffer;
    private DiffUtil.ItemCallback<E> DIFF_CALLBACK;

    public BaseAdapterPresenter() {
        this.DIFF_CALLBACK = new RAdapterAsyncDiffCallback<>();
    }

    @Override
    public void onAttach(IBaseHolderView baseHolderView) {
        this.baseHolderView = (V) baseHolderView;
    }

    @Override
    public void onAttachAdapter(BaseAdapter adapter) {
        mDiffer = new AsyncListDiffer<>(adapter, DIFF_CALLBACK);
    }

    @Override
    public void onAttachDiffCallback(DiffUtil.ItemCallback<E> callback) {
        this.DIFF_CALLBACK = callback;
    }

    @Override
    public void onAttachPayloadWatcher(RAdapterPayloadWatcher<E> payloadWatcher) {
        if (DIFF_CALLBACK instanceof RAdapterAsyncDiffCallback)
            ((RAdapterAsyncDiffCallback<E>) DIFF_CALLBACK)
                    .addPayloadWatcher(payloadWatcher);
    }

    public V view() {
        return baseHolderView;
    }

    @Override
    public void update(List<E> list) {
        mDiffer.submitList(list);
    }

    @Override
    public int getCount() {
//        return list != null ? list.size() : 0;
        return mDiffer.getCurrentList().size();
    }

    @Override
    public void addNewList(List<E> listNewItems) {
        List<E> oldList = new ArrayList<>(mDiffer.getCurrentList());
        oldList.addAll(listNewItems);
        mDiffer.submitList(oldList);
    }

    @Override
    public void addItem(E data) {
        List<E> oldList = new ArrayList<>(mDiffer.getCurrentList());
        oldList.add(data);
        mDiffer.submitList(oldList);
    }

    @Override
    public void addItemAt(int position, E data) {
        List<E> oldList = new ArrayList<>(mDiffer.getCurrentList());
        oldList.add(position, data);
        mDiffer.submitList(oldList);
    }

    @Override
    public void removeItem(int position) {
        List<E> oldList = new ArrayList<>(mDiffer.getCurrentList());
        oldList.remove(position);
        mDiffer.submitList(oldList);
    }

    @Override
    public E getFrom(int position) {
        return mDiffer.getCurrentList().get(position);
    }

    @Override
    public List<E> getAll() {
        return mDiffer.getCurrentList();
    }
}

File: SampleApp/basefiles/src/main/java/com/rp/basefiles/BaseAdapterPresenter.kt
package com.rp.basefiles

import androidx.recyclerview.widget.AsyncListDiffer
import androidx.recyclerview.widget.DiffUtil
import com.rp.util.adapter.RAdapterAsyncDiffCallback
import com.rp.util.adapter.RAdapterPayloadWatcher


abstract class BaseAdapterPresenter<V : IBaseHolderView, E> : IBaseAdapterPresenter<E> {

    private var baseHolderView: V? = null

    private var mDiffer: AsyncListDiffer<E>? = null
    private var DIFF_CALLBACK: DiffUtil.ItemCallback<E>? = null

    init {
        this.DIFF_CALLBACK = RAdapterAsyncDiffCallback<E>()
    }

    @Suppress("UNCHECKED_CAST")
    override fun onAttach(baseHolderView: IBaseHolderView) {
        this.baseHolderView = baseHolderView as V
    }

    override fun onAttachAdapter(adapter: BaseAdapter<*, *>) {
        mDiffer = AsyncListDiffer<E>(adapter, DIFF_CALLBACK!!)
    }

    override fun onAttachDiffCallback(callback: DiffUtil.ItemCallback<E>) {
        this.DIFF_CALLBACK = callback
    }

    override fun onAttachPayloadWatcher(payloadWatcher: RAdapterPayloadWatcher<E>) {
        if (DIFF_CALLBACK is RAdapterAsyncDiffCallback<E>)
            (DIFF_CALLBACK as RAdapterAsyncDiffCallback<E>)
                    .addPayloadWatcher(payloadWatcher)
    }

    fun view(): V? = baseHolderView

    override fun update(list: List<E>) {
        mDiffer?.submitList(list)
    }

    override val count: Int
        get() = mDiffer?.currentList?.size ?: 0

    override fun addNewList(listNewItems: List<E>) {

        val oldList = ArrayList<E>(mDiffer?.currentList as MutableCollection<out E>)
        oldList.addAll(listNewItems)
        mDiffer?.submitList(oldList)
    }

    override fun addItem(data: E) {
        val oldList = ArrayList<E>(mDiffer?.currentList as MutableCollection<out E>)
        oldList.add(data)
        mDiffer?.submitList(oldList)
    }

    override fun addItemAt(position: Int, data: E) {
        val oldList = ArrayList<E>(mDiffer?.currentList as MutableCollection<out E>)
        oldList.add(position, data)
        mDiffer?.submitList(oldList)
    }

    override fun removeItem(position: Int) {
        val oldList = ArrayList<E>(mDiffer?.currentList as MutableCollection<out E>)
        oldList.removeAt(position)
        mDiffer?.submitList(oldList)
    }

    override fun changeItemAt(position: Int, data: E) {
        val oldList = ArrayList<E>(mDiffer?.currentList as MutableCollection<out E>)
        oldList[position] = data
        mDiffer?.submitList(oldList)
    }

    override fun getFrom(position: Int): E? = mDiffer?.currentList?.get(position)

    override val all: List<E>
        get() = mDiffer?.currentList as List<E>
}